File: Search Ranking/google-pagerank.java
// Algorithm: PageRank
// Type: Graph, Ranking, Link Analysis
// Time Complexity: O(k * (V + E))
// Space Complexity: O(V)

import java.util.*;

public class PageRank {
    public static Map<String, Double> compute(Map<String, List<String>> graph, double d, int maxIter) {
        int N = graph.size();
        Map<String, Double> rank = new HashMap<>();
        for (String node : graph.keySet()) {
            rank.put(node, 1.0 / N);
        }

        for (int iter = 0; iter < maxIter; iter++) {
            Map<String, Double> newRank = new HashMap<>();
            for (String node : graph.keySet()) {
                newRank.put(node, (1 - d) / N);
            }

            for (String node : graph.keySet()) {
                List<String> outLinks = graph.get(node);
                if (!outLinks.isEmpty()) {
                    double share = rank.get(node) / outLinks.size();
                    for (String neighbor : outLinks) {
                        newRank.put(neighbor, newRank.get(neighbor) + d * share);
                    }
                }
            }
            rank = newRank;
        }

        return rank;
    }
}

File: Search Ranking/google-pagerank.c
// Algorithm: PageRank
// Type: Graph, Ranking, Link Analysis
// Time Complexity: O(k * (V + E))
// Space Complexity: O(V)

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NODES 100
#define DAMPING 0.85
#define MAX_ITER 100

void pagerank(int graph[MAX_NODES][MAX_NODES], int N, double rank[MAX_NODES]) {
    double new_rank[MAX_NODES];
    for (int i = 0; i < N; i++) rank[i] = 1.0 / N;

    for (int iter = 0; iter < MAX_ITER; iter++) {
        for (int i = 0; i < N; i++) new_rank[i] = (1 - DAMPING) / N;

        for (int u = 0; u < N; u++) {
            int out_deg = 0;
            for (int j = 0; j < N; j++) if (graph[u][j]) out_deg++;

            if (out_deg > 0) {
                for (int v = 0; v < N; v++) {
                    if (graph[u][v])
                        new_rank[v] += DAMPING * (rank[u] / out_deg);
                }
            }
        }
        memcpy(rank, new_rank, sizeof(double) * N);
    }
}
